src/number_theory/sum_four_squares.lean

```diff
@@ -0,0 +1,219 @@
+-/
+import data.zmod.basic field_theory.finite group_theory.perm.sign
+
+open finset polynomial finite_field equiv
+
+lemma add_even_of_add_square_even {m x y : ℤ} (h : 2 * m = x^2 + y^2) : (x + y) % 2 = 0 :=
+have hzmod : ∀ x y : zmod 2, x ^ 2 + y ^ 2 = 0 → x + y = 0, from dec_trivial,
+have (x : zmod 2)^2 + (y : zmod 2)^2 = 0,
+  by simpa [show (2 : zmod 2) = 0, from rfl, eq_comm] using congr_arg (coe : ℤ → zmod 2) h,
+show x + y ≡ 0 [ZMOD 2], from (zmod.eq_iff_modeq_int' (show 2 > 0, from dec_trivial)).1 $
+by simpa using hzmod _ _ this
+
+lemma sub_even_of_add_square_even {m x y : ℤ} (h : 2 * m = x^2 + y^2) : (x - y) % 2 = 0 :=
+add_even_of_add_square_even (show 2 * m = x^2 + (-y)^2, by simpa)
+
+private lemma sum_two_squares_of_two_mul_sum_two_squares {m x y : ℤ} (h : 2 * m = x^2 + y^2) :
+  m = ((x - y) / 2) ^ 2 + ((x + y) / 2) ^ 2 :=
+(domain.mul_left_inj (show (2*2 : ℤ) ≠ 0, from dec_trivial)).1 $
+calc 2 * 2 * m = (x - y)^2 + (x + y)^2 : by rw [mul_assoc, h]; ring
+... = (2 * ((x - y) / 2))^2 + (2 * ((x + y) / 2))^2 :
+  by rw [int.mul_div_cancel' ((int.dvd_iff_mod_eq_zero _ _).2 (sub_even_of_add_square_even h)),
+    int.mul_div_cancel' ((int.dvd_iff_mod_eq_zero _ _).2 (add_even_of_add_square_even h))]
+... = 2 * 2 * (((x - y) / 2) ^ 2 + ((x + y) / 2) ^ 2) :
+  by simp [mul_add, pow_succ, mul_comm, mul_assoc, mul_left_comm]
+
+private lemma sum_four_squares_of_two_mul_sum_four_squares {m a b c d : ℤ}
+  (h : a^2 + b^2 + c^2 + d^2 = 2 * m) : ∃ w x y z, w^2 + x^2 + y^2 + z^2 = m :=
+have ∀ f : fin 4 → zmod 2, (f 0)^2 + (f 1)^2 + (f 2)^2 + (f 3)^2 = 0 →
+    ∃ i : (fin 4), (f i)^2 + f (swap i 0 1)^2 = 0 ∧ f (swap i 0 2)^2 + f (swap i 0 3)^2 = 0,
+  from dec_trivial,
+let f : fin 4 → ℤ := vector.nth (a::b::c::d::vector.nil) in
+let ⟨i, hσ⟩ := this (coe ∘ f) (by rw [← @zero_mul (zmod 2) _ m, ← show ((2 : ℤ) : zmod 2) = 0, from rfl,
+  ← int.cast_mul, ← h]; simp only [int.cast_add, int.cast_pow]; refl) in
+let σ := swap i 0 in
+have h01 : 2 ∣ f (σ 0) ^ 2 + f (σ 1) ^ 2,
+  from (@zmod.eq_zero_iff_dvd_int 2 _).1 $ by simpa [σ] using hσ.1,
+have h23 : 2 ∣ f (σ 2) ^ 2 + f (σ 3) ^ 2,
+  from (@zmod.eq_zero_iff_dvd_int 2 _).1 $ by simpa using hσ.2,
+let ⟨x, hx⟩ := h01 in let ⟨y, hy⟩ := h23 in
+⟨(f (σ 0) - f (σ 1)) / 2, (f (σ 0) + f (σ 1)) / 2, (f (σ 2) - f (σ 3)) / 2, (f (σ 2) + f (σ 3)) / 2,
+  begin
+    rw [← sum_two_squares_of_two_mul_sum_two_squares hx.symm, add_assoc,
+      ← sum_two_squares_of_two_mul_sum_two_squares hy.symm,
+      ← domain.mul_left_inj (show (2 : ℤ) ≠ 0, from dec_trivial), ← h, mul_add, ← hx, ← hy],
+    have : univ.sum (λ x, f (σ x)^2) = univ.sum (λ x, f x^2),
+    { conv_rhs { rw finset.sum_univ_perm σ } },
+    have fin4univ : (univ : finset (fin 4)).1 = 0::1::2::3::0, from dec_trivial,
+    simpa [finset.sum_eq_multiset_sum, fin4univ, multiset.sum_cons, f]
+  end⟩
+
+lemma neg_one_sum_two_squares_mod_p {p : ℕ} (hp : p.prime) : ∃ a b : zmodp p hp, a^2 + b^2 + 1 = 0 :=
```

Comment:
As your proof shows, this works in an arbitrary field of positive characteristic. Please prove it in that generality. I'm quite sure this lemma is used more often.

Reply:
Done.